File: creational/Java/AbstractFactory.java
import java.util.ArrayList;
import java.util.List;

interface Shape {
	void draw();
}

class Circle implements Shape {

	@Override
	public void draw() {
		System.out.println("Circle: draw");
	}

}

class Ellipse implements Shape {

	@Override
	public void draw() {
		System.out.println("Ellipse: draw");
	}

}

class Square implements Shape {

	@Override
	public void draw() {
		System.out.println("Square: draw");
	}

}

class Rectangle implements Shape {

	@Override
	public void draw() {
		System.out.println("Rectangle: draw");
	}

}

interface ShapeFactory {

	Shape createCurvedInstance();

	Shape createStraightInstance();

}

class SimpleShapeFactory implements ShapeFactory {

	@Override
	public Shape createCurvedInstance() {
		return new Circle();
	}

	@Override
	public Shape createStraightInstance() {
		return new Square();
	}

}

class ComplexShapeFactory implements ShapeFactory {

	@Override
	public Shape createCurvedInstance() {
		return new Ellipse();
	}

	@Override
	public Shape createStraightInstance() {
		return new Rectangle();
	}

}

class Main {

	public static void main(String[] args) {
		System.out.println(":: Abstract Factory ::\n");

		List<Shape> shapes = new ArrayList<Shape>();

		SimpleShapeFactory simpleFactory = new SimpleShapeFactory();
		shapes.add(simpleFactory.createCurvedInstance());
		shapes.add(simpleFactory.createStraightInstance());

		ComplexShapeFactory complexFactory = new ComplexShapeFactory();
		shapes.add(complexFactory.createCurvedInstance());
		shapes.add(complexFactory.createStraightInstance());

		for (Shape shape : shapes) {
			shape.draw();
		}
	}

}

File: creational/PHP/AbstractFactory.php
<?php

interface Shape {
  public function draw();
}

class Circle implements Shape {

  public function draw() {
    echo "Circle: draw\n";
  }

}

class Ellipse implements Shape {

  public function draw() {
    echo "Ellipse: draw\n";
  }

}

class Square implements Shape {

  public function draw() {
    echo "Square: draw\n";
  }

}

class Rectangle implements Shape {

  public function draw() {
    echo "Rectangle: draw\n";
  }

}

interface ShapeFactory {

  public function createCurvedInstance();

  public function createStraightInstance();

}

class SimpleShapeFactory implements ShapeFactory {

  public function createCurvedInstance() {
    return new Circle();
  }

  public function createStraightInstance() {
    return new Square();
  }

}

class ComplexShapeFactory implements ShapeFactory {

  public function createCurvedInstance() {
    return new Ellipse();
  }

  public function createStraightInstance() {
    return new Rectangle();
  }

}

echo ":: Abstract Factory ::\n\n";

$shapes = [];

$simpleFactory = new SimpleShapeFactory();
array_push($shapes, $simpleFactory->createCurvedInstance());
array_push($shapes, $simpleFactory->createStraightInstance());

$complexFactory = new ComplexShapeFactory();
array_push($shapes, $complexFactory->createCurvedInstance());
array_push($shapes, $complexFactory->createStraightInstance());

foreach ($shapes as $shape) {
  $shape->draw();
}

?>
